File: engine/model/src/main/scala/client/BuddyListEntry.scala
package client

class BuddyListEntry(private var _name: String, var group: String, private var _characterId: Int, var channel: Int, private var _visible: Boolean) {
  def name: String = _name

  def characterId: Int = _characterId

  def visible: Boolean = _visible

  override def hashCode(): Int = {
    val prime: Int = 31
    var result: Int = 1
    result = prime * result + _characterId
    result
  }

  override def equals(obj: Any): Boolean = {
    if (this == obj) {
      return true
    }
    if (obj == null) {
      return false
    }
    if (getClass != obj.getClass) {
      return false
    }
    val other: BuddyListEntry = obj.asInstanceOf[BuddyListEntry]
    _characterId == other.characterId
  }
}

File: engine/engine-model/src/main/java/client/BuddyListEntry.java
package client;

import java.util.Objects;

public record BuddyListEntry(String name, String group, int characterId, int channel, boolean visible) {
   public BuddyListEntry updateChannel(int channel) {
      return new BuddyListEntry(name, group, characterId, channel, visible);
   }

   @Override
   public boolean equals(Object o) {
      if (this == o) {
         return true;
      }
      if (o == null || getClass() != o.getClass()) {
         return false;
      }
      BuddyListEntry that = (BuddyListEntry) o;
      return Objects.equals(characterId, that.characterId);
   }

   @Override
   public int hashCode() {
      return Objects.hash(characterId);
   }
}
